src/tvp/controller.ts

```diff
@@ -0,0 +1,79 @@
+import { commands, workspace, Disposable } from "coc.nvim"
+
+import { TreeModel } from "./model"
+import { TreeViewFeature } from "./feature"
+import { TreeViewsManager, WindowProp } from "./treeviews"
+
+export class TreeViewController implements Disposable {
+  private treeModels: Map<string, TreeModel> = new Map()
+  private listeners: Disposable[] = []
+
+  public constructor(
+    treeViewFeature: TreeViewFeature,
+    private treeViewsManager: TreeViewsManager
+  ) {
+    this.listeners.push(
+      treeViewFeature.providerEvents()(provider => {
+        const model = new TreeModel(provider)
+        this.treeModels.set(provider.viewId, model)
+        treeViewsManager.addTreeModel(model)
+      })
+    )
+
+    this.listeners.push(
+      commands.registerCommand("metals.tvp", viewName => {
+        if (viewName === undefined) {
+          treeViewsManager.toggleAllTrees()
+        } else {
+          treeViewsManager.toggleTreeView(viewName)
+        }
+      })
+    )
+
+    this.listeners.push(
+      commands.registerCommand("metals.tvp.view", (action, viewId) => {
```

Comment:
So this is also showing up in the command pallete with no description. Can we add it into the `package.json` with a title so it will show when a user sees it in the command pallete? Same as with `metals.tvp`, can we extract action into a type?

Reply:
Added.